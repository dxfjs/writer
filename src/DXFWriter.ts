import DXFManager   from "./Internals/DXFManager";
import Header       from "./Sections/Header/Header";
import Tables       from "./Sections/Tables/Tables";
import Blocks       from "./Sections/Blocks/Blocks";
import Classes      from "./Sections/Classes/Classes";
import Objects      from "./Sections/Objects/Objects";
import Entities     from "./Sections/Entities/Entities";

export default class DXFWriter extends DXFManager {
    private _header:    Header;
    private _classes:   Classes;
    private _tables:    Tables;
    private _blocks:    Blocks;
    private _entities:  Entities;
    private _objects:   Objects;

    public constructor(version: string = DXFManager.versions.R2007) {
        super();
        this.setVersion(version);
        this._header    = new Header();
        this._classes   = new Classes();
        this._tables    = new Tables();
        this._blocks    = new Blocks();
        this._entities  = new Entities();
        this._objects   = new Objects();
    }

    public addVariable(name: string, group_code: number, value: number | string): DXFWriter {
        this._header.addVariable(name, group_code, value);
        return this;
    }

    public addLineType(name: string, descriptive: string, elements: number []): DXFWriter {
        this._tables.addLineType(name, descriptive, elements);
        return this;
    }

    public addLayer(name: string, color: number, ltype: string, flag: number = 0): DXFWriter {
        this._tables.addLayer(name, color, ltype, flag);
        return this;
    }

    public setCurrentLayer(layerName: string): DXFWriter
    {
        if (this._tables.layers.find((layer) => layer.layerName === layerName)) {
            DXFManager.currentLayer = layerName;
        }
        return this;
    }

    public setUnit(unit: number): DXFWriter {
        this._header.unit = unit;
        return this;
    }

    public setVersion(version: string): DXFWriter {
        DXFManager.version = version;
        return this;
    }

    public addLine(
        x_start: number, y_start: number,
        x_end: number, y_end: number
    ): DXFWriter {
        this._entities.addLine(x_start, y_start, x_end, y_end);
        return this;
    }

    public addPolyline(points: number[][], flag: number): DXFWriter {
        this._entities.addPolyline(points, flag);
        return this;
    }

    public addRectangle (
        top_left_x: number, top_left_y: number,
        bottom_right_x: number, bottom_right_y: number
    ): DXFWriter {
        const corners = [
            [top_left_x,        top_left_y],
            [bottom_right_x,    top_left_y],
            [bottom_right_x,    bottom_right_y],
            [top_left_x,        bottom_right_y]
        ];
        this._entities.addPolyline(corners, 1);
        return this;
    }

    public addPolyline3D(points: number[][], flag: number): DXFWriter {
        this._entities.addPolyline3D(points, flag);
        return this;
    }

    public addPoint(x: number, y: number, z: number): DXFWriter {
        this._entities.addPoint(x, y, z);
        return this;
    }

    public addCircle(x_center: number, y_center: number, radius: number): DXFWriter {
        this._entities.addCircle(x_center, y_center, radius);
        return this;
    }

    public addArc(
        x: number, y: number, radius: number,
        startAngle: number, endAngle: number
    ): DXFWriter {
        this._entities.addArc(x, y, radius, startAngle, endAngle);
        return this;
    }

    public addSpline(
        controlPoints: number[][], fitPoints: number[][], curveDegree: number,
        flag: number, knots: number[],
        weights: number[]
    ): DXFWriter {
        this._entities.addSpline(controlPoints, fitPoints, curveDegree, flag, knots, weights);
        return this;
    }

    public addEllipse(
        x_center: number, y_center: number, x_major_axis: number, y_major_axis: number,
        ratio_minor_axis: number, start_parameter: number, end_parameter: number
    ): DXFWriter {
        this._entities.addEllipse(
            x_center, y_center, x_major_axis, y_major_axis, ratio_minor_axis,
            start_parameter, end_parameter
        );
        return this;
    }

    public add3DFace(
        x_first: number, y_first: number, z_first: number,
        x_second: number, y_second: number, z_second: number,
        x_third: number, y_third: number, z_third: number,
        x_fourth: number, y_fourth: number, z_fourth: number
    ): DXFWriter {
        this._entities.add3DFace(x_first, y_first, z_first,
            x_second, y_second, z_second,
            x_third, y_third, z_third,
            x_fourth, y_fourth, z_fourth
        );
        return this;
    }

    public addText(
        x: number, y: number,
        height: number, value: string,
    ): DXFWriter {
        this._entities.addText(
            x, y, height, value
        );
        return this;
    }

    public stringify(): string {
        let str: string = '';
        str += this._header.stringify();
        str += this._classes.stringify();
        str += this._tables.stringify();
        this._blocks.modelHandle = this._tables.blockRecords.modelHandle; // TODO remove this from here
        this._blocks.paperHandle = this._tables.blockRecords.paperHandle; // TODO remove this from here
        str += this._blocks.stringify();
        str += this._entities.stringify();
        str += this._objects.stringify();
        return str;
    }
}