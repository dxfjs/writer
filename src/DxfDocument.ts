import GlobalState from './GlobalState';
import Handle from './Internals/Handle';
import DxfInterface from './Internals/Interfaces/DxfInterface';
import TagsManager, { point3d_t } from './Internals/TagsManager';
import DxfBlock from './Sections/BlocksSection/DxfBlock';
import DxfBlocksSection from './Sections/BlocksSection/DxfBlocksSection';
import DxfClassesSection from './Sections/ClassesSection/DxfClassesSection';
import DxfEntitiesSection from './Sections/EntitiesSection/DxfEntitiesSection';
import Image, {
	ImageOptions_t,
} from './Sections/EntitiesSection/Entities/Image';
import DxfHeaderSection from './Sections/HeaderSection/DxfHeaderSection';
import DxfObjects from './Sections/ObjectsSection/DxfObjectsSection';
import DxfImageDef from './Sections/ObjectsSection/Objects/DxfImageDef';
import DxfImageDefReactor from './Sections/ObjectsSection/Objects/DxfImageDefReactor';
import DxfTablesSection from './Sections/TablesSection/DxfTablesSection';
import DxfVPort from './Sections/TablesSection/Tables/Records/DxfVPort';

export default class DxfDocument implements DxfInterface {
	readonly header: DxfHeaderSection;
	readonly classes: DxfClassesSection;
	readonly tables: DxfTablesSection;
	readonly blocks: DxfBlocksSection;
	readonly entities: DxfEntitiesSection;
	readonly objects: DxfObjects;
	readonly activeVPort: DxfVPort;
	readonly modelSpace: DxfBlock;
	readonly paperSpace: DxfBlock;

	constructor() {
		this.header = new DxfHeaderSection();
		this.classes = new DxfClassesSection();
		this.tables = new DxfTablesSection();
		this.blocks = new DxfBlocksSection(this.tables);
		this.entities = new DxfEntitiesSection(this.blocks.modelSpace);
		this.objects = new DxfObjects();

		this.header.setVariable('$ACADVER', { 1: 'AC1021' });
		this.handseed();
		this.header.setVariable('$INSUNITS', { 70: GlobalState.units });

		this.tables.addLType('ByBlock', '', []);
		this.tables.addLType('ByLayer', '', []);
		this.tables.addLType('Continuous', 'Solid line', []);
		this.tables.addLayer('0', 7, 'Continuous', 0);
		this.tables.addStyle('Standard');
		this.tables.addAppId('ACAD', 0);
		this.tables.addDimStyle('Standard');
		this.activeVPort = this.tables.addVPort('*Active');

		this.modelSpace = this.blocks.modelSpace;
		this.paperSpace = this.blocks.paperSpace;
	}

	addBlock(name: string) {
		return this.blocks.addBlock(name);
	}

	setCurrentLayerName(name: string): void {
		const layerRecord = this.tables.layerTable.layerRecords.find(
			(layer) => layer.name === name
		);
		if (layerRecord) GlobalState.currentLayerName = name;
		else throw new Error(`The '${name} layer doesn't exist!'`);
	}

	private handseed() {
		this.header.setVariable('$HANDSEED', { 5: Handle.peek() });
	}

	setUnits(units: number) {
		GlobalState.units = units;
		this.header.setVariable('$INSUNITS', { 70: GlobalState.units });
	}

	setViewCenter(center: point3d_t) {
		this.header.setVariable('$VIEWCTR', {
			10: center.x,
			20: center.y,
		});
		this.activeVPort.viewCenter = [center.x, center.y];
	}

	addImage(
		imagePath: string,
		name: string,
		insertionPoint: point3d_t,
		width: number,
		height: number,
		scale: number,
		rotation: number,
		options?: ImageOptions_t
	): Image {
		// TODO make sure there is no IMAGEDEF for this image!
		const imageDef = new DxfImageDef(imagePath);
		imageDef.width = width;
		imageDef.height = height;
		const image = new Image(
			{
				height,
				width,
				scale,
				rotation,
				insertionPoint,
				imageDefHandle: imageDef.handle,
			},
			options
		);
		const imageDefReactor = new DxfImageDefReactor(image.handle);
		image.imageDefReactorHandle = imageDefReactor.handle;
		this.modelSpace.addEntity(image);
		this.objects.addObject(imageDef);
		this.objects.addObject(imageDefReactor);
		const dictionary = this.objects.addDictionary();

		dictionary.addEntryObject(name, imageDef.handle);
		imageDef.ownerObjecthandle = dictionary.handle;
		this.objects.root.addEntryObject('ACAD_IMAGE_DICT', dictionary.handle);
		imageDef.acadImageDictHandle = dictionary.handle;
		imageDef.addImageDefReactorHandle(imageDefReactor.handle);
		return image;
	}

	stringify(): string {
		this.handseed();
		this.setViewCenter(this.modelSpace.centerView()); // fit in
		this.activeVPort.viewHeight = this.modelSpace.viewHeight();
		return this.manager.stringify();
	}

	get manager(): TagsManager {
		const manager = new TagsManager();
		manager.append(this.header);
		manager.append(this.classes);
		manager.append(this.tables);
		manager.append(this.blocks);
		manager.append(this.entities);
		manager.append(this.objects);
		return manager;
	}
}
