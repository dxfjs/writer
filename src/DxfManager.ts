import GlobalState from './GlobalState';
import Handle from './Internals/Handle';
import DxfInterface from './Internals/Interfaces/DxfInterface';
import TagsManager, { point3d_t } from './Internals/TagsManager';
import DxfBlock from './Sections/BlocksSection/DxfBlock';
import DxfBlocksSection from './Sections/BlocksSection/DxfBlocksSection';
import DxfClassesSection from './Sections/ClassesSection/DxfClassesSection';
import DxfEntitiesSection from './Sections/EntitiesSection/DxfEntitiesSection';
import Image, {
	ImageOptions_t,
} from './Sections/EntitiesSection/Entities/Image';
import DxfHeaderSection from './Sections/HeaderSection/DxfHeaderSection';
import DxfObjects from './Sections/ObjectsSection/DxfObjectsSection';
import DxfImageDef from './Sections/ObjectsSection/Objects/DxfImageDef';
import DxfImageDefReactor from './Sections/ObjectsSection/Objects/DxfImageDefReactor';
import DxfTablesSection from './Sections/TablesSection/DxfTablesSection';
import DxfViewPort from './Sections/TablesSection/Tables/Records/DxfViewPort';

export default class DxfManager implements DxfInterface {
	static #instance: DxfManager;
	readonly headerSection: DxfHeaderSection;
	readonly classesSection: DxfClassesSection;
	readonly tablesSection: DxfTablesSection;
	readonly blocksSection: DxfBlocksSection;
	readonly entitiesSection: DxfEntitiesSection;
	readonly objectsSection: DxfObjects;
	readonly activeViewPort: DxfViewPort;
	readonly modelSpace: DxfBlock;
	readonly paperSpace: DxfBlock;

	private constructor() {
		this.headerSection = DxfHeaderSection.getInstance();
		this.classesSection = DxfClassesSection.getInstance();
		this.tablesSection = DxfTablesSection.getInstance();
		this.blocksSection = DxfBlocksSection.getInstance();
		this.entitiesSection = DxfEntitiesSection.getInstance();
		this.objectsSection = DxfObjects.getInstance();

		this.headerSection.setVariable('$ACADVER', { 1: 'AC1021' });
		this.updateHandle();
		this.headerSection.setVariable('$INSUNITS', { 70: GlobalState.units });

		this.tablesSection.addLineType('ByBlock', '', []);
		this.tablesSection.addLineType('ByLayer', '', []);
		this.tablesSection.addLineType('Continuous', 'Solid line', []);
		this.tablesSection.addLayer('0', 7, 'Continuous', 0);
		this.tablesSection.addStyle('Standard');
		this.tablesSection.addAppId('ACAD', 0);
		this.tablesSection.addDimStyle('Standard');
		this.activeViewPort = this.tablesSection.addViewPort('*Active');

		this.modelSpace = this.blocksSection.modelSpace;
		this.paperSpace = this.blocksSection.paperSpace;
	}

	static getInstance(): DxfManager {
		if (!this.#instance) this.#instance = new DxfManager();
		return this.#instance;
	}

	addBlock(name: string) {
		return this.blocksSection.addBlock(name);
	}

	setCurrentLayerName(name: string): void {
		const layerRecord = this.tablesSection.layerTable.layerRecords.find(
			(layer) => layer.name === name
		);
		if (layerRecord) GlobalState.currentLayerName = name;
		else throw new Error(`The '${name} layer doesn't exist!'`);
	}

	private updateHandle() {
		this.headerSection.setVariable('$HANDSEED', { 5: Handle.peek() });
	}

	setUnits(units: number) {
		GlobalState.units = units;
		this.headerSection.setVariable('$INSUNITS', { 70: GlobalState.units });
	}

	setViewCenter(center: point3d_t) {
		this.headerSection.setVariable('$VIEWCTR', {
			10: center.x,
			20: center.y,
		});
		this.activeViewPort.viewCenter = [center.x, center.y];
	}

	addImage(
		imagePath: string,
		name: string,
		insertionPoint: point3d_t,
		width: number,
		height: number,
		scale: number,
		rotation: number,
		options?: ImageOptions_t
	): Image {
		// TODO make sure there is no IMAGEDEF for this image!
		const imageDef = new DxfImageDef(imagePath);
		imageDef.width = width;
		imageDef.height = height;
		const image = new Image(
			{
				height,
				width,
				scale,
				rotation,
				insertionPoint,
				imageDefHandle: imageDef.handle,
			},
			options
		);
		const imageDefReactor = new DxfImageDefReactor(image.handle);
		image.imageDefReactorHandle = imageDefReactor.handle;
		this.modelSpace.addEntity(image);
		this.objectsSection.addObject(imageDef);
		this.objectsSection.addObject(imageDefReactor);
		const dictionary = this.objectsSection.addDictionary();

		dictionary.addEntryObject(name, imageDef.handle);
		imageDef.ownerObjecthandle = dictionary.handle;
		this.objectsSection.rootDictionary.addEntryObject(
			'ACAD_IMAGE_DICT',
			dictionary.handle
		);
		imageDef.acadImageDicId = dictionary.handle;
		imageDef.addImageDefReactorId(imageDefReactor.handle);
		return image;
	}

	stringify(): string {
		this.updateHandle();
		this.setViewCenter(this.modelSpace.centerView()); // fit in
		this.activeViewPort.viewHeight = this.modelSpace.viewHeight();
		return this.manager.stringify();
	}

	get manager(): TagsManager {
		const manager = new TagsManager();
		manager.append(this.headerSection);
		manager.append(this.classesSection);
		manager.append(this.tablesSection);
		manager.append(this.blocksSection);
		manager.append(this.entitiesSection);
		manager.append(this.objectsSection);
		return manager;
	}
}
