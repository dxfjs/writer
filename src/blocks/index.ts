export * from "./block";
export * from "./blocks";
export * from "./endblk";
