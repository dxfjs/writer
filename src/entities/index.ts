export * from "./arc";
export * from "./circle";
export * from "./dimension";
export * from "./ellipse";
export * from "./entities";
export * from "./entity";
export * from "./face";
export * from "./hatch";
export * from "./insert";
export * from "./leader";
export * from "./line";
export * from "./lwpolyline";
export * from "./manager";
export * from "./mesh";
export * from "./mleader";
export * from "./mtext";
export * from "./point";
export * from "./polyline";
export * from "./ray";
export * from "./seqend";
export * from "./solid";
export * from "./spline";
export * from "./table";
export * from "./text";
export * from "./vertex";
