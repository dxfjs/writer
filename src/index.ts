export * from './DxfWriter';
export * from './Sections/HeaderSection/DxfVariable';
export * from './Internals/TagsManager';
export * from './Sections/EntitiesSection/Entity';
export * from './Sections/TablesSection/Tables/Records/DxfRecord';
export * from './Sections/EntitiesSection/Entities/LWPolyline';
export * from './Internals/Utils';
export * from './Sections/EntitiesSection/Entities/Spline';
export * from './Sections/BlocksSection/DxfBlock';
export * from './Sections/TablesSection/Tables/Records/DxfAppId';
export * from './Sections/TablesSection/Tables/Records/DxfBlockRecord';
export * from './Sections/TablesSection/Tables/Records/DxfDimStyle';
export * from './Sections/TablesSection/Tables/Records/DxfLType';
export * from './Sections/TablesSection/Tables/Records/DxfLayer';
export * from './Sections/ObjectsSection/Objects/DxfImageDef';
export * from './Sections/EntitiesSection/Entities/Face';
export * from './Sections/EntitiesSection/Entities/Image';
export * from './Sections/EntitiesSection/Entities/Insert';
export * from './Internals/HatchPatterns';
export * from './Sections/EntitiesSection/Entities/Hatch';
export * from './Sections/ObjectsSection/Objects/DxfImageDef';
export * from './Sections/ObjectsSection/Objects/DxfImageDefReactor';
export * from './Internals/Enums';
