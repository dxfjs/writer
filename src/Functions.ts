import DxfManager from './DxfManager';
import {
	HatchBoundaryPath,
	HatchGradientOptions_t,
	HatchOptions_t,
	HatchPatternOptions_t,
} from './Sections/EntitiesSection/Entities/Hatch';
import { point3d_t } from './Internals/TagsManager';
import DxfBlock from './Sections/BlocksSection/DxfBlock';
import DxfBlocksSection from './Sections/BlocksSection/DxfBlocksSection';
import DxfEntitiesSection from './Sections/EntitiesSection/DxfEntitiesSection';
import Arc from './Sections/EntitiesSection/Entities/Arc';
import Circle from './Sections/EntitiesSection/Entities/Circle';
import Face, { faceOptions_t } from './Sections/EntitiesSection/Entities/Face';
import { ImageOptions_t } from './Sections/EntitiesSection/Entities/Image';
import Insert, {
	insertOptions_t,
} from './Sections/EntitiesSection/Entities/Insert';
import Line from './Sections/EntitiesSection/Entities/Line';
import LWPolyline, {
	lwPolylineOptions_t,
	lwPolylineVertex_t,
} from './Sections/EntitiesSection/Entities/LWPolyline';
import { options_t } from './Sections/EntitiesSection/Entity';
import DxfObject from './Sections/ObjectsSection/DxfObject';
import DxfObjects from './Sections/ObjectsSection/DxfObjectsSection';
import DxfDictionary from './Sections/ObjectsSection/Objects/DxfDictionary';
import DxfImageDef from './Sections/ObjectsSection/Objects/DxfImageDef';
import DxfImageDefReactor from './Sections/ObjectsSection/Objects/DxfImageDefReactor';
import DxfTablesSection from './Sections/TablesSection/DxfTablesSection';
import DxfAppId from './Sections/TablesSection/Tables/Records/DxfAppId';
import DxfBlockRecord from './Sections/TablesSection/Tables/Records/DxfBlockRecord';
import DxfDimStyle from './Sections/TablesSection/Tables/Records/DxfDimStyle';
import DxfLayer from './Sections/TablesSection/Tables/Records/DxfLayer';
import DxfLineType from './Sections/TablesSection/Tables/Records/DxfLineType';
import DxfStyle from './Sections/TablesSection/Tables/Records/DxfStyle';

//#region Tables
/**
 * @public
 * @param name -
 * @returns
 */
export function addBlock(name: string): DxfBlock {
	return DxfBlocksSection.getInstance().addBlock(name);
}

/**
 * @public
 * @param name -
 * @param flags -
 * @returns
 */
export function addAppId(name: string, flags?: number): DxfAppId {
	return DxfTablesSection.getInstance().addAppId(name, flags);
}

/**
 * @public
 * @param name -
 * @returns
 */
export function addBlockRecord(name: string): DxfBlockRecord {
	return DxfTablesSection.getInstance().addBlockRecord(name);
}

/**
 * @public
 * @param name -
 * @param flags -
 * @returns
 */
export function addDimStyle(name: string, flags?: number): DxfDimStyle {
	return DxfTablesSection.getInstance().addDimStyle(name, flags);
}

/**
 * @public
 * @param name -
 * @param descriptive -
 * @param elements -
 * @param flags -
 * @returns
 */
export function addLineType(
	name: string,
	descriptive: string,
	elements: number[],
	flags?: number
): DxfLineType {
	return DxfTablesSection.getInstance().addLineType(
		name,
		descriptive,
		elements,
		flags
	);
}

/**
 * @public
 * @param name -
 * @param color -
 * @param lineType -
 * @param flags -
 * @returns
 */
export function addLayer(
	name: string,
	color: number,
	lineType: string,
	flags?: number
): DxfLayer {
	return DxfTablesSection.getInstance().addLayer(
		name,
		color,
		lineType,
		flags
	);
}

/**
 * @public
 * @param name - Tha name of style.
 * @returns
 */
export function addStyle(name: string): DxfStyle {
	return DxfTablesSection.getInstance().addStyle(name);
}
//#endregion

//#region Entities
/**
 *
 * @public
 *
 * @param center -
 * @param radius -
 * @param startAngle -
 * @param endAngle -
 * @param options -
 * @returns
 */
export function addArc(
	center: point3d_t,
	radius: number,
	startAngle: number,
	endAngle: number,
	options?: options_t
): Arc {
	return DxfEntitiesSection.getInstance().modelSpace.addArc(
		center,
		radius,
		startAngle,
		endAngle,
		options
	);
}

/**
 *
 * @param center
 * @param radius
 * @param options
 * @returns
 */
export function addCircle(
	center: point3d_t,
	radius: number,
	options?: options_t
): Circle {
	return DxfEntitiesSection.getInstance().modelSpace.addCircle(
		center,
		radius,
		options
	);
}

/**
 *
 * @param center
 * @param endPointOfMajorAxis
 * @param ratioOfMinorAxisToMajorAxis
 * @param startParameter
 * @param endParameter
 * @param options
 * @returns
 */
export function addEllipse(
	center: point3d_t,
	endPointOfMajorAxis: point3d_t,
	ratioOfMinorAxisToMajorAxis: number,
	startParameter: number,
	endParameter: number,
	options?: options_t
) {
	return DxfEntitiesSection.getInstance().modelSpace.addEllipse(
		center,
		endPointOfMajorAxis,
		ratioOfMinorAxisToMajorAxis,
		startParameter,
		endParameter,
		options
	);
}

/**
 *
 * @param firstCorner
 * @param secondCorner
 * @param thirdCorner
 * @param fourthCorner
 * @param options
 * @returns
 */
export function add3dFace(
	firstCorner: point3d_t,
	secondCorner: point3d_t,
	thirdCorner: point3d_t,
	fourthCorner: point3d_t,
	options?: faceOptions_t
): Face {
	return DxfEntitiesSection.getInstance().modelSpace.add3dFace(
		firstCorner,
		secondCorner,
		thirdCorner,
		fourthCorner,
		options
	);
}

/**
 *
 * @param absolutePath
 * @param name
 * @param insertionPoint
 * @param width
 * @param height
 * @param scale
 * @param rotation
 * @param options
 * @returns
 */
export function addImage(
	absolutePath: string,
	name: string,
	insertionPoint: point3d_t,
	width: number,
	height: number,
	scale: number,
	rotation: number,
	options?: ImageOptions_t
) {
	return DxfManager.getInstance().addImage(
		absolutePath,
		name,
		insertionPoint,
		width,
		height,
		scale,
		rotation,
		options
	);
}

/**
 * Add an insert entity to the Dxf.
 * @public
 * @param blockName - The name of the block to insert.
 * @param insertionPoint - The point where the block is to be inserted.
 * @param options - The options of the Insert entity.
 * @returns Return a reference to the added Insert entity.
 */
export function addInsert(
	blockName: string,
	insertionPoint: point3d_t,
	options?: insertOptions_t
): Insert {
	return DxfEntitiesSection.getInstance().modelSpace.addInsert(
		blockName,
		insertionPoint,
		options
	);
}

/**
 *
 * @param startPoint
 * @param endPoint
 * @param options
 * @returns
 */
export function addLine(
	startPoint: point3d_t,
	endPoint: point3d_t,
	options?: options_t
): Line {
	return DxfEntitiesSection.getInstance().modelSpace.addLine(
		startPoint,
		endPoint,
		options
	);
}

/**
 *
 * @param points
 * @param options
 * @returns
 */
export function addLWPolyline(
	points: lwPolylineVertex_t[],
	options: lwPolylineOptions_t = {}
): LWPolyline {
	return DxfEntitiesSection.getInstance().modelSpace.addLWPolyline(
		points,
		options
	);
}

/**
 *
 * @param boundaryPath
 * @param fill
 * @param options
 * @returns
 */
export function addHatch(
	boundaryPath: HatchBoundaryPath,
	fill: HatchPatternOptions_t | HatchGradientOptions_t,
	options?: HatchOptions_t
) {
	return DxfEntitiesSection.getInstance().modelSpace.addHatch(
		boundaryPath,
		fill,
		options
	);
}

//#endregion

//#region Objects

/**
 * Add an object to objects section.
 * @public
 * @param object
 * @returns
 */
export function addObject<T extends DxfObject>(object: T): T {
	return DxfObjects.getInstance().addObject(object);
}

/**
 * @public
 * @returns
 */
export function addDictionary(): DxfDictionary {
	return DxfObjects.getInstance().addDictionary();
}

/**
 * @public
 * @param name
 * @param softOwner
 */
export function addEntryToRootDictionary(
	name: string,
	softOwner: string
): void {
	DxfObjects.getInstance().rootDictionary.addEntryObject(name, softOwner);
}

/**
 * @public
 * @param path
 * @returns
 */
export function addImageDef(path: string): DxfImageDef {
	return addObject(new DxfImageDef(path));
}

/**
 * @public
 * @param dictionaryId
 * @returns
 */
export function addImageDefReactor(dictionaryId: string) {
	return addObject(new DxfImageDefReactor(dictionaryId));
}

//#endregion
