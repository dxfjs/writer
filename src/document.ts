import { BlockOptions, XBlocks } from "./blocks";
import { Units, XBBox, XHandle, XTagsManager, point2d } from "./utils";
import { Stringifiable } from "./types";
import { XClasses } from "./classes";
import { XEntities } from "./entities";
import { XHeader } from "./header";
import { XObjects } from "./objects";
import { XTables } from "./tables";

export class XDocument implements Stringifiable {
  readonly handle: XHandle;
  readonly header: XHeader;
  readonly classes: XClasses;
  readonly blocks: XBlocks;
  readonly entities: XEntities;
  readonly tables: XTables;
  readonly objects: XObjects;

  units: number;

  get modelSpace() {
    return this.blocks.modelSpace;
  }

  get paperSpace() {
    return this.blocks.paperSpace;
  }

  constructor() {
    this.handle = new XHandle();
    this.header = new XHeader(this.handle);
    this.classes = new XClasses();
    this.tables = new XTables(this.handle);
    this.blocks = new XBlocks(this.tables, this.handle);
    this.entities = new XEntities(this.blocks);
    this.objects = new XObjects(this.handle);

    this.units = Units.Unitless;
  }

  setUnits(units: number) {
    this.units = units;
  }

  addBlock(options: BlockOptions) {
    this.tables.addBlockRecord(options);
    return this.blocks.addBlock(options);
  }

  addPaperSpace() {
    return this.blocks.addPaperSpace();
  }

  addVariable(name: string) {
    return this.header.add(name);
  }

  stringify(): string {
    const mg = new XTagsManager();
    this.fitIn();
    this.header.tagify(mg);
    this.classes.tagify(mg);
    this.tables.tagify(mg);
    this.blocks.tagify(mg);
    this.entities.tagify(mg);
    this.objects.tagify(mg);
    mg.add(0, "EOF");
    return mg.stringify();
  }

  private fitIn() {
    const bbox = this.modelSpace.bbox();
    const center = XBBox.center(bbox);
    const height = XBBox.height(bbox);
    this.tables.vportActive.lowerLeft = point2d(bbox.minX, bbox.minY);
    this.tables.vportActive.upperRight = point2d(bbox.maxX, bbox.maxY);
    this.tables.vportActive.center = center;
    this.tables.vportActive.height = height;
  }
}
