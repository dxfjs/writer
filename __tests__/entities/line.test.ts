import { XHandle, XLine, XTagsManager, point } from "../../src";

describe("XLine class", () => {
  it("should create a line entity", () => {
    const line = new XLine(
      {
        start: point(),
        end: point(100, 100),
      },
      new XHandle()
    );
    const mg = new XTagsManager();
    line.tagify(mg);
    expect(mg.stringify()).toMatchSnapshot();
  });
});
