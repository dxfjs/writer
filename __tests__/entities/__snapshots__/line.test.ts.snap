// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`XLine class > should create a line entity 1`] = `
"0
LINE
5
1
330
0
100
AcDbEntity
67
0
8
ByLayer
6
ByLayer
347
ByLayer
62
256
370
-3
60
0
100
AcDbLine
39
0
10
0
20
0
30
0
11
100
21
100
31
0
210
0
220
0
230
1"
`;
